teuthology/task/ansible.py

```diff
@@ -0,0 +1,239 @@
+from tempfile import NamedTemporaryFile
+
+from teuthology.exceptions import CommandFailedError
+from teuthology.repo_utils import fetch_repo
+
+from . import Task
+
+log = logging.getLogger(__name__)
+
+
+class LoggerFile(object):
+    """
+    A thin wrapper around a logging.Logger instance that provides a file-like
+    interface.
+
+    Used by Ansible.execute_playbook() when it calls pexpect.run()
+    """
+    def __init__(self, logger, level):
+        self.logger = logger
+        self.level = level
+
+    def write(self, string):
+        self.logger.log(self.level, string)
+
+    def flush(self):
+        pass
+
+
+class Ansible(Task):
+    """
+    A task to run ansible playbooks
+
+    Required configuration parameters:
+        playbook:   Required; can either be a list of plays, or a path/URL to a
+                    playbook
+
+    Optional configuration parameters:
+        repo:       A path or URL to a repo (defaults to '.'). Given a repo
+                    value of 'foo', ANSIBLE_ROLES_PATH is set to 'foo/roles'
+        branch:     If pointing to a remote git repo, use this branch. Defaults
+                    to 'master'.
+        hosts:      A list of teuthology roles or partial hostnames (or a
+                    combination of the two). ansible-playbook will only be run
+                    against hosts that match.
+        inventory:  A path to be passed to ansible-playbook with the
+                    --inventory-file flag; useful for playbooks that also have
+                    vars they need access to. If this is not set, we check for
+                    /etc/ansible/hosts and use that if it exists. If it does
+                    not, we generate a temporary file to use.
+        tags:       A string including any (comma-separated) tags to be passed
+                    directly to ansible-playbook.
+
+    Examples:
+
+    tasks:
+    - ansible:
+        repo: https://github.com/ceph/ceph-cm-ansible.git
+        playbook:
+          - roles:
+            - some_role
+            - another_role
+        hosts:
+          - client.0
+          - host1
+
+    tasks:
+    - ansible:
+        repo: /path/to/repo
+        inventory: /path/to/inventory
+        playbook: /path/to/playbook.yml
+        tags: my_tags
+
+    """
+    def __init__(self, ctx, config):
+        super(Ansible, self).__init__(ctx, config)
+        self.log = log
+        self.generated_inventory = False
+        self.generated_playbook = False
+
+    def setup(self):
+        super(Ansible, self).setup()
+        self.find_repo()
+        self.get_playbook()
+        self.get_inventory() or self.generate_hosts_file()
+        if not hasattr(self, 'playbook_file'):
+            self.generate_playbook()
+
+    def find_repo(self):
+        """
+        Locate the repo we're using; cloning it from a remote repo if necessary
+        """
+        repo = self.config.get('repo', '.')
+        is_remote = (repo.startswith('http://') or
```

Comment:
`startswith` can take a tuple of strings to try, instead of doing `or` multiple times:

    is_remote = repo.startswith(('http', 'git', 'https'))